ref-exchange/src/stable_swap/mod.rs

```diff
@@ -0,0 +1,526 @@
+        self.total_fee
+    }
+
+    /// Returns volumes of the given pool.
+    pub fn get_volumes(&self) -> Vec<SwapVolume> {
+        self.volumes.clone()
+    }
+
+    /// Add liquidity into the pool.
+    /// Allows to add liquidity of a subset of tokens.
+    pub fn add_liquidity(
+        &mut self,
+        sender_id: &AccountId,
+        amounts: &mut Vec<Balance>,
+        fees: &SwapFees,
+    ) -> Balance {
+        assert_eq!(
+            amounts.len(),
+            self.token_account_ids.len(),
+            "ERR_WRONG_TOKEN_COUNT"
+        );
+        let invariant = StableSwap::new(
+            self.init_amp_factor,
+            self.target_amp_factor,
+            env::block_timestamp(),
+            self.init_amp_time,
+            self.stop_amp_time,
+        );
+        let new_shares = if self.shares_total_supply == 0 {
+            // Bootstrapping the pool.
+            invariant
+                .compute_d(amounts[0], amounts[1])
+                .expect("ERR_CALC_FAILED")
+                .as_u128()
+        } else {
+            invariant
+                .compute_lp_amount_for_deposit(
+                    amounts[0],
+                    amounts[1],
+                    self.amounts[0],
+                    self.amounts[1],
+                    self.shares_total_supply,
+                    &Fees::new(self.total_fee, &fees),
+                )
+                // TODO: proper error
+                .expect("ERR_CALC_FAILED")
+        };
+
+        // TODO: add slippage check on the LP tokens.
+        self.amounts[0] += amounts[0];
+        self.amounts[1] += amounts[1];
+
+        self.mint_shares(sender_id, new_shares);
+        new_shares
+    }
+
+    /// Remove liquidity from the pool.
+    /// Allows to remove liquidity of a subset of tokens, by providing 0 in `min_amount` for the tokens to not withdraw.
+    pub fn remove_liquidity(
+        &mut self,
+        sender_id: &AccountId,
+        shares: Balance,
+        min_amounts: Vec<Balance>,
+        fees: &SwapFees,
+    ) -> Vec<Balance> {
+        assert_eq!(
+            min_amounts.len(),
+            self.token_account_ids.len(),
+            "ERR_WRONG_TOKEN_COUNT"
+        );
+        let prev_shares_amount = self.shares.get(&sender_id).expect("ERR_NO_SHARES");
+        assert!(prev_shares_amount >= shares, "ERR_NOT_ENOUGH_SHARES");
+        let mut result = vec![0u128; N_COINS as usize];
+        let invariant = StableSwap::new(
+            self.init_amp_factor,
+            self.target_amp_factor,
+            env::block_timestamp(),
+            self.init_amp_time,
+            self.stop_amp_time,
+        );
+        let mut fee_amounts = vec![0u128; N_COINS as usize];
+        let stable_swap_fees = Fees::new(self.total_fee, &fees);
+        for (idx, min_amount) in min_amounts.iter().enumerate() {
+            if *min_amount != 0 {
+                let (amount_out, fee) = invariant
+                    .compute_withdraw_one(
```

Comment:
It seems `compute_withdraw_one` has figured out the outcome for remove `shares` with one type of token. When user want withdraw both tokens,  each output token balance represent the shares, so cause an insistency with share reduced and tokens balances withdrawed.